configmerge/configmerge.go

```diff
@@ -0,0 +1,244 @@
+package configmerge
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/bitrise-io/bitrise/models"
+	"github.com/bitrise-io/go-utils/sliceutil"
+	logV2 "github.com/bitrise-io/go-utils/v2/log"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	MaxIncludeCountPerFile = 10
+	MaxFilesCountTotal     = 20
+	MaxIncludeDepth        = 5           // root + 4 includes
+	MaxFileSizeBytes       = 1024 * 1024 // 1MB
+)
+
+func IsModularConfig(mainConfigPth string) (bool, error) {
+	mainConfigFile, err := os.Open(mainConfigPth)
+	if err != nil {
+		return false, err
+	}
+	mainConfigContent, err := io.ReadAll(mainConfigFile)
+	if err != nil {
+		return false, err
+	}
+
+	var config struct {
+		Include []ConfigReference `yaml:"include" json:"include"`
+	}
+	if err := yaml.Unmarshal(mainConfigContent, &config); err != nil {
+		return false, err
+	}
+	return len(config.Include) > 0, nil
+}
+
+type RepoInfoProvider interface {
+	GetRepoInfo(repoPth string) (*RepoInfo, error)
+}
+
+type FileReader interface {
+	ReadFileFromFileSystem(name string) ([]byte, error)
+	ReadFileFromGitRepository(repository string, branch string, commit string, tag string, path string) ([]byte, error)
+}
+
+type Merger struct {
+	repoInfoProvider RepoInfoProvider
+	fileReader       FileReader
+	logger           logV2.Logger
+
+	repoInfo *RepoInfo
+
+	filesCount int
+}
+
+func NewMerger(repoInfoProvider RepoInfoProvider, fileReader FileReader, logger logV2.Logger) Merger {
+	return Merger{
+		repoInfoProvider: repoInfoProvider,
+		fileReader:       fileReader,
+		logger:           logger,
+	}
+}
+
+func (m *Merger) MergeConfig(mainConfigPth string) (string, *models.ConfigFileTreeModel, error) {
```

Comment:
This `Merger` struct maybe could be a bit more higher level. It does not necessarily need to know if a config comes from a local file or remote one. What it needs to do is to read and understand the include section of the yamls, then ask an entity to return the bytes of the next yaml file, build a tree out of these and then merge it. 

The other parts could be abstracted away a bit so you only know about them when you start digging deeper.

Reply:
Created a new ConfigReader from the original FileReader and moved the local vs repo file read logic there.